kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,1338 @@
+    kernel_type: &DataType,
+    arrow_type: &ArrowDataType,
+) -> DeltaResult<DataTypeCompat> {
+    match (kernel_type, arrow_type) {
+        (DataType::Primitive(_), _) if arrow_type.is_primitive() => {
+            let converted_type: ArrowDataType = kernel_type.try_into()?;
+            if &converted_type == arrow_type {
+                Ok(DataTypeCompat::Identical)
+            } else {
+                check_cast_compat(arrow_type, converted_type)
+            }
+        }
+        (DataType::Primitive(PrimitiveType::Boolean), ArrowDataType::Boolean)
+        | (DataType::Primitive(PrimitiveType::String), ArrowDataType::Utf8)
+        | (DataType::Primitive(PrimitiveType::Binary), ArrowDataType::Binary) => {
+            // strings, bools, and binary  aren't primitive in arrow
+            Ok(DataTypeCompat::Identical)
+        }
+        (
+            DataType::Primitive(PrimitiveType::Decimal(kernel_prec, kernel_scale)),
+            ArrowDataType::Decimal128(arrow_prec, arrow_scale),
+        ) if arrow_prec == kernel_prec && *arrow_scale == *kernel_scale as i8 => {
+            // decimal isn't primitive in arrow. cast above is okay as we limit range
+            Ok(DataTypeCompat::Identical)
+        }
+        (DataType::Array(inner_type), ArrowDataType::List(arrow_list_type)) => {
+            let kernel_array_type = &inner_type.element_type;
+            let arrow_list_type = arrow_list_type.data_type();
+            ensure_data_types(kernel_array_type, arrow_list_type)
+        }
+        (DataType::Map(kernel_map_type), ArrowDataType::Map(arrow_map_type, _)) => {
+            if let ArrowDataType::Struct(fields) = arrow_map_type.data_type() {
+                let mut fiter = fields.iter();
+                if let Some(key_type) = fiter.next() {
+                    ensure_data_types(&kernel_map_type.key_type, key_type.data_type())?;
+                } else {
+                    return Err(make_arrow_error(
+                        "Arrow map struct didn't have a key type".to_string(),
+                    ));
+                }
+                if let Some(value_type) = fiter.next() {
+                    ensure_data_types(&kernel_map_type.value_type, value_type.data_type())?;
+                } else {
+                    return Err(make_arrow_error(
+                        "Arrow map struct didn't have a value type".to_string(),
+                    ));
+                }
+                Ok(DataTypeCompat::Nested)
+            } else {
+                Err(make_arrow_error(
+                    "Arrow map type wasn't a struct.".to_string(),
+                ))
+            }
+        }
+        (DataType::Struct(kernel_fields), ArrowDataType::Struct(arrow_fields)) => {
+            // build a list of kernel fields that matches the order of the arrow fields
+            let mapped_fields = arrow_fields
+                .iter()
+                .flat_map(|f| kernel_fields.fields.get(f.name()));
```

Comment:
nit: [filter_map](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map) is probably clearer than `flat_map` here? We're specifically trying keep the content of `Some` entries corresponding to successful name lookups, right?